src/leap/mail/adaptors/soledad.py

```diff
@@ -0,0 +1,1047 @@
+        :param fdoc: a dictionary containing values from which a
+                     FlagsDocWrapper can be initialized
+        :type fdoc: dict
+        :param hdoc: a dictionary containing values from which a
+                     HeaderDocWrapper can be initialized
+        :type hdoc: dict
+        :param cdocs: None, or a dictionary mapping integers (1-indexed) to
+                      dicts from where a ContentDocWrapper can be initialized.
+        :type cdocs: dict, or None
+
+        :rtype: MessageClass instance.
+        """
+        assert(MessageClass is not None)
+        return MessageClass(MessageWrapper(mdoc, fdoc, hdoc, cdocs), uid=uid)
+
+    def _get_msg_from_variable_doc_list(self, doc_list, msg_class, uid=None):
+        if len(doc_list) == 3:
+            mdoc, fdoc, hdoc = doc_list
+            cdocs = None
+        elif len(doc_list) > 3:
+            fdoc, hdoc = doc_list[:3]
+            cdocs = dict(enumerate(doc_list[3:], 1))
+        return self.get_msg_from_docs(
+            msg_class, mdoc, fdoc, hdoc, cdocs, uid=None)
+
+    def get_msg_from_mdoc_id(self, MessageClass, store, mdoc_id,
+                             uid=None, get_cdocs=False):
+
+        def wrap_meta_doc(doc):
+            cls = MetaMsgDocWrapper
+            return cls(doc_id=doc.doc_id, **doc.content)
+
+        def get_part_docs_from_mdoc_wrapper(wrapper):
+            d_docs = []
+            d_docs.append(store.get_doc(wrapper.fdoc))
+            d_docs.append(store.get_doc(wrapper.hdoc))
+            for cdoc in wrapper.cdocs:
+                d_docs.append(store.get_doc(cdoc))
+
+            def add_mdoc(doc_list):
+                return [wrapper.serialize()] + doc_list
+
+            d = defer.gatherResults(d_docs)
+            d.addCallback(add_mdoc)
+            return d
+
+        def get_parts_doc_from_mdoc_id():
+            mbox = re.findall(constants.METAMSGID_MBOX_RE, mdoc_id)[0]
+            chash = re.findall(constants.METAMSGID_CHASH_RE, mdoc_id)[0]
+
+            def _get_fdoc_id_from_mdoc_id():
+                return constants.FDOCID.format(mbox_uuid=mbox, chash=chash)
+
+            def _get_hdoc_id_from_mdoc_id():
+                return constants.HDOCID.format(mbox_uuid=mbox, chash=chash)
+
+            d_docs = []
+            fdoc_id = _get_fdoc_id_from_mdoc_id()
+            hdoc_id = _get_hdoc_id_from_mdoc_id()
+
+            d_docs.append(store.get_doc(mdoc_id))
+            d_docs.append(store.get_doc(fdoc_id))
+            d_docs.append(store.get_doc(hdoc_id))
+
+            d = defer.gatherResults(d_docs)
+            return d
+
+        if get_cdocs:
+            d = store.get_doc(mdoc_id)
+            d.addCallback(wrap_meta_doc)
+            d.addCallback(get_part_docs_from_mdoc_wrapper)
+        else:
+            d = get_parts_doc_from_mdoc_id()
+
+        d.addCallback(partial(self._get_msg_from_variable_doc_list,
+                              msg_class=MessageClass, uid=uid))
```

Comment:
true! this must be something that evolved from a different use, or maybe a lapsus.